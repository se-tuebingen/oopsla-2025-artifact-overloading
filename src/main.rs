use overloading::compile_expr;
use overloading::syntax::parse_expr;

use clap::Parser;

#[derive(Parser)]
#[command(name = "overloading")]
struct Args {
    /// Input file to process
    filename: String,

    /// Use short output format
    #[arg(short, long)]
    short: bool,
}

/// Main entry point for the native CLI.
fn main() {
    let args = Args::parse();

    let content = std::fs::read_to_string(&args.filename).expect("Failed to read file");
    match parse_expr(&content) {
        Ok(expr) => {
            println!("{}", compile_expr(&expr, args.short));
        }
        Err(err) => {
            panic!("Failed to parse expression: {}", err);
        }
    }
}

/// Snapshot tests for the compiler examples.
#[cfg(test)]
#[test]
fn test_compiler_examples() {
    use insta::{assert_json_snapshot, glob, with_settings};

    glob!("../", "examples/*.over", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let output = match parse_expr(&input) {
            Ok(expr) => overloading::compile_expr_json(&expr, true),
            Err(err) => overloading::json::JsonProcessingResult::InferenceError {
                expression: input,
                error: err,
            },
        };

        // Use the filename (without extension) as snapshot suffix
        let filename = path.file_stem().unwrap().to_str().unwrap();
        with_settings!({
            snapshot_path => "../examples/snapshots",
            snapshot_suffix => filename,
            sort_maps => true,
        }, {
            assert_json_snapshot!(output);
        });
    });
}
