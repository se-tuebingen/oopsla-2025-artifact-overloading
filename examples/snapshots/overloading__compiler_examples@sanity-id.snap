---
source: src/main.rs
expression: output
input_file: examples/sanity-id.over
---
{
  "status": "success",
  "expression": "assume idPoly: ∀(T ). (T) => T\nassume idInt: (Int) => Int\nassume addInt: (Int, Int) => Int\nassume addString: (String, String) => String\nλ x => b‹addInt, addString›(x, a‹idPoly, idInt›(42)) => \n  b‹addInt, addString›(x, a‹idPoly, idInt›(42))",
  "dimension_spans": "[spans]",
  "inferred_type": "(x_0) => app_5",
  "worlds": {
    "type": "many",
    "count": 2.0
  },
  "solutions": [
    [
      "b₁",
      "a₁"
    ],
    [
      "b₁",
      "a₂"
    ]
  ],
  "constraints": [
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "b₁"
      ],
      "to": "over_b_1"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "b₂"
      ],
      "to": "over_b_1"
    },
    {
      "from": "(T_3) => T_3",
      "choices": [
        "a₁"
      ],
      "to": "over_a_2"
    },
    {
      "from": "(Int) => Int",
      "choices": [
        "a₂"
      ],
      "to": "over_a_2"
    },
    {
      "from": "over_a_2",
      "choices": [],
      "to": "(Int) => app_4"
    },
    {
      "from": "over_b_1",
      "choices": [],
      "to": "(x_0, app_4) => app_5"
    }
  ],
  "bounds": [
    {
      "variable": "T_3",
      "lower": [
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "app_4",
          "choices": [
            "a₁"
          ]
        }
      ]
    },
    {
      "variable": "app_4",
      "lower": [
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "a₂"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "Int",
          "choices": [
            "b₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "b₂"
          ]
        }
      ]
    },
    {
      "variable": "app_5",
      "lower": [
        {
          "type_name": "Int",
          "choices": [
            "b₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "b₂"
          ]
        }
      ],
      "upper": []
    },
    {
      "variable": "over_a_2",
      "lower": [
        {
          "type_name": "(Int) => Int",
          "choices": [
            "a₂"
          ]
        },
        {
          "type_name": "(T_3) => T_3",
          "choices": [
            "a₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(Int) => app_4",
          "choices": []
        }
      ]
    },
    {
      "variable": "over_b_1",
      "lower": [
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "b₁"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "b₂"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(x_0, app_4) => app_5",
          "choices": []
        }
      ]
    },
    {
      "variable": "x_0",
      "lower": [],
      "upper": [
        {
          "type_name": "Int",
          "choices": [
            "b₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "b₂"
          ]
        }
      ]
    }
  ],
  "errors": [
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "b₂",
          "a₁"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "b₂",
          "a₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    }
  ]
}
